File: crypto_agent/tools/cancel_order.py
import json
from .registry import register_tool


@register_tool(
    name="cancel_order",
    description="Cancel an open order by ID.",
    schema={
        "type": "object",
        "properties": {
            "order_id": {"type": "string", "description": "Order ID to cancel"},
            "symbol": {"type": "string", "description": "Trading pair the order belongs to"},
        },
        "required": ["order_id", "symbol"],
    },
)
async def handle_cancel_order(exchange, order_id: str, symbol: str, **_) -> str:
    try:
        result = await exchange.cancel_order(order_id, symbol)
        return json.dumps(result, indent=2)
    except Exception as e:
        return f"Error: {e}"

File: src/tools/cancel-order.ts
import { registerTool } from "./registry.js";

registerTool(
  "cancel_order",
  "Cancel an open order by ID.",
  {
    type: "object",
    properties: {
      order_id: { type: "string", description: "Order ID to cancel" },
      symbol: { type: "string", description: "Trading pair the order belongs to" },
    },
    required: ["order_id", "symbol"],
  },
  ["exchange", "broker", "config"],
  async ({ exchange, broker, config, order_id, symbol }) => {
    try {
      if (config?.paperTrading) {
        if (!broker) return "Error: paper broker is not initialized";
        const result = await broker.cancelOrder(order_id, symbol);
        return JSON.stringify(result, null, 2);
      }
      if (!exchange) return "Error: live exchange is not initialized";
      const result = await exchange.cancelOrder(order_id, symbol);
      return JSON.stringify(result, null, 2);
    } catch (e: any) {
      return `Error: ${e.message ?? e}`;
    }
  },
);
